### PR TITLE
Applab theme padding

### PR DESCRIPTION
**Pull request by cpirich**
_Context_

* Applab elements have always had some padding, but the values were usually specified in CSS outside of the applab scope. The new theme comps include consistent padding of `15px` on the sides so that the text isn't hugging against the edge of the element.

_Considerations_
* Introducing padding as a student configurable property on elements is not desirable at this time. To do it properly would require four separate properties, which would introduce a lot of complexity in the design mode editor.
* Changing the existing padding values for all projects will create compatibility issues

_Design Plan_
* Introduce padding as an internal-only property that can be set/retrieved only as part of changing themes. The syntax of the padding value will match the CSS spec, so we can maintain a single padding string value.
* Preserve (and explicitly backfill) the historical padding values for each element as part of the `Classic` theme, which ensures that we don't introduce any compatibility issues. Students will not need to deal with truncated text or resizing elements until they try changing themes (still currently under an experiment).
* Set the same padding values for all of our new themes. Once the themes experiment is permanently enabled, all new projects will create screens in the `Default` theme and all elements placed on the screen will have the new padding values. Changing to any theme other than `Classic` will preserve the new padding values.
* The `label` element now has padding on the left and right sides, even when the label has a transparent background. This consistency makes it easier to align `label` text above `dropdown`, `textInput`, and `textArea` elements. It also simplifies theme changes, ensuring that the label text stays in the same left-aligned position.

_Bonus updates to label resizing_
* The `label` element has some special magic to resize automatically if the element was deemed to be sized exactly to match the text. I've updated this logic to account for `padding` changes and the recently introduced `fontFamily` property. I've also added the notion of a `batchChangeId` during `updateProperty()` calls. This optional parameter is set when property values are changed in batches during theme changes. It allows the `label` to be smart about measuring just once at the beginning of a batch of a property updates (to determine if the element was previously "sized to fit").

_Example screenshots (note: buttons are not impacted by padding changes)_
Screen Shot 2019-05-01 at 11 42 18 AM
Screen Shot 2019-05-01 at 11 42 40 AM

### CODE REVIEW

**Comment by cpirich** [comment]
@ryansloan FYI

File: apps/src/applab/designElements/elementUtils.js [hunk 1]
```diff
@@ -227,3 +227,56 @@ export function setDefaultBorderStyles(element, options = {}) {
...
   }
 }
+
+/**
+ * Parse a padding string and return the total horizontal padding and
+ * total vertical padding.
+ * @param {string} cssPaddingString value from element.style.padding
+ */
+export function calculatePadding(cssPaddingString) {
+  const paddingRegEx = /\s*?(\d*)(px)?\s*(\d*)?(px)?\s*(\d*)?(px)?\s*(\d*)?(px)?\s*?/i;
```
Comment (ajpal) [inline apps/src/applab/designElements/elementUtils.js:237]: Can you add a comment explaining what this regex matches?

File: apps/src/applab/designElements/elementUtils.js [hunk 2]
```diff
@@ -227,3 +227,56 @@ export function setDefaultBorderStyles(element, options = {}) {
...
+  const paddingResult = paddingRegEx.exec(cssPaddingString);
+
+  const paddingValues = [
+    parseInt(paddingResult[1], 10),
```
Comment (ajpal) [inline apps/src/applab/designElements/elementUtils.js:241]: Comments for these as well? left, right, top, bottom?

File: apps/src/applab/designElements/label.jsx [hunk 3]
```diff
@@ -400,19 +422,41 @@ export default {
     element.style.height = size.height + 'px';
   },
 
+  _lastFitsExactly: {},
```
Comment (ajpal) [inline apps/src/applab/designElements/label.jsx:425]: What is this for? Could you add a comment or two explaining this and what batchChangeId is for?

File: apps/src/applab/designElements/label.jsx [hunk 4]
```diff
@@ -400,19 +422,41 @@ export default {
...
+  beforePropertyChange: function(element, name, batchChangeId) {
+    switch (name) {
+      case 'padding':
+      case 'text':
+      case 'fontFamily':
+      case 'fontSize': {
+        const {
+          batchId = -1,
+          previouslyFitExactly: batchPreviouslyFitExactly
+        } = this._lastFitsExactly;
```
Comment (ajpal) [inline apps/src/applab/designElements/label.jsx:439]: What does this do? I haven't seen this syntax before

**Review by ajpal — APPROVED** [review]
lgtm, pending a few requests for comments.

**Comment by cpirich** [comment]
Thanks @ajpal - will include requested comments in my follow-up PR.

File: apps/src/applab/designElements/label.jsx (hunk 4 above)
Comment (cpirich) [reply apps/src/applab/designElements/label.jsx:439]: This is object destructuring from ES6. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring

**Comment by joshlory** [comment]
Follow up to PR https://github.com/code-dot-org/code-dot-org/pull/27914.

**Pull request merged**